src/MSBuildLocator/DotNetSdkLocationHelper.cs

```diff
@@ -71,45 +71,101 @@ internal static class DotNetSdkLocationHelper
+                RemoveUnmanagedDllResolver();
+            }
+
+            Dictionary<Version, VisualStudioInstance?> versionInstanceMap = new();
+            foreach (var basePath in allAvailableSdks)
             {
                 var dotnetSdk = GetInstance(basePath, allowQueryAllRuntimes);
                 if (dotnetSdk != null)
                 {
-                    yield return dotnetSdk;
+                    // We want to return the best SDK first
+                    if (dotnetSdk.VisualStudioRootPath == bestSdkPath)
+                    {
+                        // We will add a null entry to the map to ensure we do not add the same SDK from a different location.
+                        versionInstanceMap[dotnetSdk.Version] = null;
+                        yield return dotnetSdk;
+                    }
+
+                    // Only add an SDK once, even if it's installed in multiple locations.
+                    if (!versionInstanceMap.ContainsKey(dotnetSdk.Version))
+                    {
+                        versionInstanceMap.Add(dotnetSdk.Version, dotnetSdk);
+                    }
                 }
             }
-        }
 
-        private static IEnumerable<string> GetDotNetBasePaths(string workingDirectory)
-        {
-            try
+            // We want to return the newest SDKs first. Using OfType will remove the null entry added if we found the best SDK.
+            var instances = versionInstanceMap.Values.OfType<VisualStudioInstance>().OrderByDescending(i => i.Version);
+            foreach (var instance in instances)
             {
-                AddUnmanagedDllResolver();
+                yield return instance;
+            }
 
-                string? bestSDK = GetSdkFromGlobalSettings(workingDirectory);
-                if (!string.IsNullOrEmpty(bestSDK))
+            // Returns the list of all available SDKs ordered by ascending version.
+            static IEnumerable<string> GetAllAvailableSDKs()
+            {
+                bool foundSdks = false;
+                string[]? resolvedPaths = null;
+                foreach (string dotnetPath in s_dotnetPathCandidates.Value)
                 {
-                    yield return bestSDK;
-                }
+                    int rc = NativeMethods.hostfxr_get_available_sdks(exe_dir: dotnetPath, result: (key, value) => resolvedPaths = value);
 
-                string[] dotnetPaths = GetAllAvailableSDKs();
-                // We want to return the newest SDKs first, however, so iterate over the list in reverse order.
-                // If basePath is disqualified because it was later
-                // than the runtime version, this ensures that RegisterDefaults will return the latest valid
-                // SDK instead of the earliest installed.
-                for (int i = dotnetPaths.Length - 1; i >= 0; i--)
-                {
-                    if (dotnetPaths[i] != bestSDK)
+                    if (rc == 0 && resolvedPaths != null)
```

Comment:
Previous behavior was that we would not throw if `resolvedPaths` was an empty array.